File: postman/sdks/Specs SDKs/Orders/python/src/orders_sdk/net/request_chain/handlers/retry_handler.py
import random
import re

from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
from typing import Generator, Optional, Tuple
from time import sleep
from .base_handler import BaseHandler
from ...transport.request import Request
from ...transport.response import Response
from ...transport.api_error import ApiConnectionError, ApiError, ApiTimeoutError
from ...transport.request_error import RequestError


class RetryHandler(BaseHandler):
    """
    Handler for retrying requests.
    Supports configurable retry attempts, exponential backoff with jitter, and specific status codes or HTTP methods to retry.

    :ivar int _max_attempts: The maximum number of retry attempts.
    :ivar int _delay_in_milliseconds: The initial delay between retry attempts in milliseconds.
    :ivar int _max_delay_in_milliseconds: The maximum delay between retry attempts in milliseconds (caps exponential backoff).
    :ivar float _backoff_factor: Multiplier for exponential backoff.
    :ivar int _jitter_in_milliseconds: Maximum random jitter in milliseconds to add to retry delays.
    :ivar set[int] | None _status_codes_to_retry: Specific HTTP status codes to retry (None triggers default: 5xx + 408, 429).
    :ivar set[str] _http_methods_to_retry: HTTP methods that are allowed to be retried.
    """

    def __init__(self):
        """
        Initialize a new instance of RetryHandler.
        """
        super().__init__()
        self._max_attempts = 3
        self._delay_in_milliseconds = 150
        self._max_delay_in_milliseconds = 5000
        self._max_retry_after_delay_in_milliseconds = 60000
        self._backoff_factor = 2
        self._jitter_in_milliseconds = 50
        self._status_codes_to_retry = None
        self._http_methods_to_retry = {
            m.upper()
            for m in ["GET", "POST", "PUT", "DELETE", "PATCH", "HEAD", "OPTIONS"]
        }

    def handle(
        self, request: Request
    ) -> Tuple[Optional[Response], Optional[RequestError]]:
        """
        Retry the request based on configured retry settings.
        Implements exponential backoff with optional jitter between retry attempts.
        Per-request retry config (via request.config['retry']) overrides SDK defaults.

        :param Request request: The request to retry.
        :return: The response and any error that occurred.
        :rtype: Tuple[Optional[Response], Optional[RequestError]]
        :raises RequestError: If the handler chain is incomplete.
        """
        if self._next_handler is None:
            raise RequestError("Handler chain is incomplete")

        retry_config = (request.config or {}).get("retry") or {}
        max_attempts = retry_config.get("attempts", self._max_attempts)

        response, error = self._next_handler.handle(request)

        try_count = 0
        while try_count < max_attempts and self._should_retry(error, request):
            self._delay(try_count, request, error)
            response, error = self._next_handler.handle(request)
            try_count += 1

        return response, error

    def stream(
        self, request: Request
    ) -> Generator[Tuple[Optional[Response], Optional[RequestError]], None, None]:
        """
        Retry the request based on configured retry settings.
        Implements exponential backoff with optional jitter between retry attempts.
        Per-request retry config (via request.config['retry']) overrides SDK defaults.

        :param Request request: The request to retry.
        :return: The response and any error that occurred.
        :rtype: Generator[Tuple[Optional[Response], Optional[RequestError]], None, None]
        :raises RequestError: If the handler chain is incomplete.
        """
        if self._next_handler is None:
            raise RequestError("Handler chain is incomplete")

        try:
            retry_config = (request.config or {}).get("retry") or {}
            max_attempts = retry_config.get("attempts", self._max_attempts)
            try_count = 0
            stream = self._next_handler.stream(request)
            while True:
                response, error = next(stream)
                if try_count < max_attempts and self._should_retry(error, request):
                    self._delay(try_count, request, error)
                    try_count += 1
                    stream = self._next_handler.stream(request)  # Retry the request
                elif try_count >= max_attempts:
                    yield response, error
                    break
                else:
                    yield response, error

        except StopIteration:
            pass

    def _delay(
        self, try_count: int, request: Request, error: Optional[ApiError] = None
    ) -> None:
        """
        Calculate and apply delay before next retry attempt.
        A server rate-limit timing header (Retry-After / X-RateLimit-Reset) on the error,
        when present, overrides the computed exponential backoff; otherwise exponential
        backoff (capped at max delay, with optional jitter) is used.
        Per-request retry config overrides SDK defaults.

        :param int try_count: Current retry attempt number (0-indexed).
        :param Request request: The request being retried, used to read per-request retry config.
        :param Optional[ApiError] error: The error that triggered the retry (carries headers).
        """
        retry_config = (request.config or {}).get("retry") or {}
        max_retry_after = retry_config.get(
            "max_retry_after_delay_ms", self._max_retry_after_delay_in_milliseconds
        )
        header_delay_ms = self._retry_after_delay(error, max_retry_after)
        if header_delay_ms is not None:
            sleep(header_delay_ms / 1000)
            return

        base_delay = retry_config.get("delay_ms", self._delay_in_milliseconds)
        max_delay = retry_config.get("max_delay_ms", self._max_delay_in_milliseconds)
        backoff_factor = retry_config.get("backoff_factor", self._backoff_factor)
        jitter = retry_config.get("jitter_ms", self._jitter_in_milliseconds)

        # Calculate exponential backoff: initialDelay * (backoffFactor ^ try_count)
        delay = base_delay * (backoff_factor**try_count)

        # Cap at max delay
        delay = min(delay, max_delay)

        # Add jitter: random value between 0 and jitter_ms
        if jitter > 0:
            delay += random.uniform(0, jitter)

        # Convert to seconds and sleep
        sleep(delay / 1000)

    def _retry_after_delay(
        self, error: Optional[ApiError], max_retry_after_ms: float
    ) -> Optional[float]:
        """
        Return the server-directed retry delay (in milliseconds) from rate-limit response
        headers, honoring Retry-After (delta-seconds or HTTP-date) and, when absent,
        X-RateLimit-Reset (epoch seconds), clamped to max_retry_after_ms. Returns None when
        no usable header is present so the caller falls back to exponential backoff.

        :param Optional[ApiError] error: The error that triggered the retry.
        :param float max_retry_after_ms: Upper bound for a server-directed delay.
        :return: The delay in milliseconds, or None to use exponential backoff.
        :rtype: Optional[float]
        """
        headers = getattr(getattr(error, "response", None), "headers", None)
        if headers is None or max_retry_after_ms <= 0:
            return None

        # retry-after-ms (milliseconds) is a non-standard but finer-grained hint some APIs
        # send (e.g. OpenAI); it takes precedence over the whole-second Retry-After.
        retry_after_ms = headers.get("retry-after-ms")
        if retry_after_ms is not None and re.fullmatch(
            r"\d+(\.\d+)?", str(retry_after_ms).strip()
        ):
            return min(float(str(retry_after_ms).strip()), max_retry_after_ms)

        seconds = self._parse_retry_after(headers.get("Retry-After"))
        if seconds is not None:
            return min(max(seconds * 1000, 0), max_retry_after_ms)

        # X-RateLimit-Reset is interpreted as epoch seconds (the common convention).
        reset = headers.get("X-RateLimit-Reset")
        if reset is not None and str(reset).strip() != "":
            try:
                epoch = float(reset)
            except (TypeError, ValueError):
                epoch = None
            if epoch is not None:
                delta_ms = epoch * 1000 - datetime.now(timezone.utc).timestamp() * 1000
                if delta_ms > 0:
                    return min(delta_ms, max_retry_after_ms)

        return None

    def _parse_retry_after(self, value: Optional[str]) -> Optional[float]:
        """
        Parse a Retry-After header value: an integer/float number of seconds, or an
        HTTP-date (a past date yields 0). Returns the delay in seconds, or None if the
        value is empty or unparseable.

        :param Optional[str] value: The raw Retry-After header value.
        :return: The delay in seconds, or None.
        :rtype: Optional[float]
        """
        if value is None:
            return None
        trimmed = str(value).strip()
        if trimmed == "":
            return None
        if re.fullmatch(r"\d+(\.\d+)?", trimmed):
            return float(trimmed)
        # Parsing and the timestamp math both run under the guard: a pathological but
        # well-formed HTTP-date (e.g. year 9999) can raise OverflowError/OSError from
        # timestamp(), and a malformed value can raise beyond TypeError/ValueError. A bad
        # server header must never crash the retry layer, so treat any failure as "no delay".
        try:
            parsed = parsedate_to_datetime(trimmed)
            if parsed is None:
                return None
            if parsed.tzinfo is None:
                parsed = parsed.replace(tzinfo=timezone.utc)
            delta = parsed.timestamp() - datetime.now(timezone.utc).timestamp()
        except (TypeError, ValueError, OverflowError, OSError):
            return None
        return delta if delta > 0 else 0.0

    def _should_retry(self, error: Optional[ApiError], request: Request) -> bool:
        """
        Determine whether the request should be retried based on status code and HTTP method.
        By default, retries all 5xx server errors and specific 4xx client errors (408 Timeout, 429 Rate Limit).
        Per-request retry config overrides SDK defaults for status codes and HTTP methods.
        A statusless transport failure is decided by its type rather than by a status code: an
        ApiTimeoutError is retryable on a retryable method, an ApiConnectionError is not.

        :param Optional[ApiError] error: The error from the previous handler.
        :param Request request: The request being retried.
        :return: True if the request should be retried, False otherwise.
        :rtype: bool
        """
        if not error:
            return False

        retry_config = (request.config or {}).get("retry") or {}

        # Check if HTTP method is retryable. Evaluated BEFORE the status gate so the transport
        # cases below can reuse it: a non-idempotent request must not be replayed just because it
        # failed at the transport level rather than with a status.
        request_http_methods = retry_config.get("http_methods_to_retry")
        if request_http_methods is not None:
            should_retry_method = request.method.upper() in {
                m.upper() for m in request_http_methods
            }
        else:
            should_retry_method = request.method.upper() in self._http_methods_to_retry

        # FSDK-1585: a timeout carries NO status, so none of the status rules below can speak
        # about it — and `None >= 500` raises TypeError rather than returning False. A timeout
        # stays retryable: it used to arrive as a synthetic 408, which sat in the default retry
        # set, and removing that fake status must not silently stop retrying timeouts. It is
        # retryable regardless of `status_codes_to_retry`, because a list of status codes cannot
        # express an opinion about a failure that never produced one.
        if isinstance(error, ApiTimeoutError):
            return should_retry_method

        # FSDK-1573: a refused connection is NOT retried. It used to escape the handler chain as a
        # raw requests exception, so it never reached this loop at all — wrapping it in an SDK type
        # is the ticket's scope, and turning it into a retryable failure would be a behaviour
        # change nothing asked for. Checked explicitly rather than left to fall through, because
        # the status rules below would raise TypeError on its absent status.
        if isinstance(error, ApiConnectionError):
            return False

        # Check if status code is retryable
        request_status_codes = retry_config.get("status_codes_to_retry")
        if request_status_codes is not None:
            should_retry_status = error.status in request_status_codes
        elif self._status_codes_to_retry is not None:
            should_retry_status = error.status in self._status_codes_to_retry
        else:
            # Default: retry 5xx, 408 (Timeout), 429 (Rate Limit)
            should_retry_status = (
                error.status >= 500 or error.status == 408 or error.status == 429
            )

        return should_retry_status and should_retry_method

File: postman/sdks/Specs SDKs/Orders/cli/sdk/internal/clients/rest/handlers/retry_handler.go
package handlers

import (
	"errors"
	"math"
	"math/rand"
	"net/http"
	"regexp"
	"strconv"
	"strings"
	"time"

	"example.com/orders/sdk/internal/clients/rest/httptransport"
)

// rng is seeded at package init to ensure non-deterministic jitter across Go versions.
// The global math/rand source is only auto-seeded from Go 1.20 onward.
var rng = rand.New(rand.NewSource(time.Now().UnixNano()))

// calculateDelay returns the backoff duration before the next retry attempt.
// Uses exponential backoff capped at maxDelay, plus optional random jitter.
func calculateDelay(attempt int, base, max, jitter time.Duration, factor float64) time.Duration {
	delay := float64(base) * math.Pow(factor, float64(attempt))
	if delay > float64(max) {
		delay = float64(max)
	}
	if jitter > 0 {
		delay += float64(rng.Int63n(int64(jitter)))
	}
	return time.Duration(delay)
}

var retryAfterDeltaRe = regexp.MustCompile(`^\d+(\.\d+)?$`)

// retryAfterDelay returns the server-directed retry delay from rate-limit response headers,
// honoring Retry-After (delta-seconds or HTTP-date) and, when absent, X-RateLimit-Reset
// (epoch seconds), clamped to maxCap. The bool is false when no usable header is present so
// the caller falls back to the computed exponential backoff.
func retryAfterDelay(retryAfterMs, retryAfter, reset string, maxCap time.Duration) (time.Duration, bool) {
	if maxCap <= 0 {
		return 0, false
	}
	// retry-after-ms (milliseconds) is a non-standard but finer-grained hint some APIs send
	// (e.g. OpenAI); it takes precedence over the whole-second Retry-After. Uses the same
	// strict delta format as Retry-After so every SDK honors the same set of values.
	if retryAfterMs = strings.TrimSpace(retryAfterMs); retryAfterDeltaRe.MatchString(retryAfterMs) {
		if ms, err := strconv.ParseFloat(retryAfterMs, 64); err == nil {
			ns := ms * float64(time.Millisecond)
			// Guard the float->Duration conversion: a huge value clamps to the cap
			// (matching the other headers and SDKs) rather than falling through.
			if ns >= float64(math.MaxInt64) {
				return maxCap, true
			}
			return clampDelay(time.Duration(ns), maxCap), true
		}
	}
	if d, ok := parseRetryAfter(retryAfter); ok {
		return clampDelay(d, maxCap), true
	}
	// X-RateLimit-Reset (epoch seconds) is only consulted when Retry-After is absent.
	// An already-elapsed reset window is treated as stale (fall back to backoff), whereas
	// an elapsed Retry-After above resolves to 0 ("retry now") — this mirrors the Ruby SDK.
	if reset = strings.TrimSpace(reset); reset != "" {
		if epoch, err := strconv.ParseInt(reset, 10, 64); err == nil {
			if d := time.Until(time.Unix(epoch, 0)); d > 0 {
				return clampDelay(d, maxCap), true
			}
		}
	}
	return 0, false
}

// parseRetryAfter parses a Retry-After header value: integer/float delta-seconds, or an
// HTTP-date (a past date yields 0). The bool is false when the value is empty or unparseable.
func parseRetryAfter(value string) (time.Duration, bool) {
	value = strings.TrimSpace(value)
	if value == "" {
		return 0, false
	}
	if retryAfterDeltaRe.MatchString(value) {
		secs, err := strconv.ParseFloat(value, 64)
		if err != nil {
			return 0, false
		}
		ns := secs * float64(time.Second)
		// Guard the float->Duration conversion: an out-of-range value is
		// implementation-defined (negative on amd64), which would defeat clampDelay.
		if ns >= float64(math.MaxInt64) {
			return time.Duration(math.MaxInt64), true
		}
		return time.Duration(ns), true
	}
	if t, err := http.ParseTime(value); err == nil {
		if d := time.Until(t); d > 0 {
			return d, true
		}
		return 0, true
	}
	return 0, false
}

// clampDelay bounds d to [0, maxCap].
func clampDelay(d, maxCap time.Duration) time.Duration {
	if d < 0 {
		return 0
	}
	if d > maxCap {
		return maxCap
	}
	return d
}

// RetryHandler automatically retries failed requests with configurable backoff.
// Retries HTTP 5xx errors and 408/429 by default; specific status codes can be configured.
// Non-HTTP errors (network, serialization) are not retried. T is the response type, E is the error type.
type RetryHandler[T any, E any] struct {
	nextHandler Handler[T, E]
}

// NewRetryHandler creates a new retry handler. Retry configuration is read from the
// request's Config on each call, allowing per-call overrides via RequestOption.
func NewRetryHandler[T any, E any]() *RetryHandler[T, E] {
	return &RetryHandler[T, E]{}
}

// shouldRetry determines if a failed request should be retried.
// Only HTTP errors on retryable methods with retryable status codes trigger a retry.
// Non-HTTP errors (network failures, serialization errors) are not retried.
func (h *RetryHandler[T, E]) shouldRetry(errResp *httptransport.ErrorResponse[E], method string, httpMethods []string, statusCodes []int) bool {
	if !errResp.IsHTTPError {
		return false
	}

	methodAllowed := false
	for _, m := range httpMethods {
		if m == method {
			methodAllowed = true
			break
		}
	}
	if !methodAllowed {
		return false
	}

	if len(statusCodes) > 0 {
		for _, code := range statusCodes {
			if code == errResp.StatusCode {
				return true
			}
		}
		return false
	}
	return errResp.StatusCode >= 500 || errResp.StatusCode == 408 || errResp.StatusCode == 429
}

// Handle processes a request with automatic retry logic on retryable HTTP failures.
// Uses exponential backoff with jitter between attempts. Returns the first successful response
// or the final error after all attempts are exhausted.
func (h *RetryHandler[T, E]) Handle(request httptransport.Request) (*httptransport.Response[T], *httptransport.ErrorResponse[E]) {
	if h.nextHandler == nil {
		return nil, httptransport.NewErrorResponse[E](errors.New("Handler chain terminated without terminating handler"), nil)
	}

	retryConfig := request.Config.Retry
	if retryConfig.MaxAttempts <= 0 {
		retryConfig.MaxAttempts = 1
	}
	for attempt := 0; attempt < retryConfig.MaxAttempts; attempt++ {
		resp, err := h.nextHandler.Handle(request.Clone())
		if err == nil {
			return resp, nil
		}

		if !h.shouldRetry(err, request.Method, retryConfig.HTTPMethodsToRetry, retryConfig.HTTPCodesToRetry) {
			return nil, err
		}

		if attempt == retryConfig.MaxAttempts-1 {
			return nil, err
		}

		if d, ok := retryAfterDelay(err.GetHeader(http.CanonicalHeaderKey("Retry-After-Ms")), err.GetHeader(http.CanonicalHeaderKey("Retry-After")), err.GetHeader(http.CanonicalHeaderKey("X-RateLimit-Reset")), retryConfig.MaxRetryAfterDelay); ok {
			time.Sleep(d)
		} else {
			time.Sleep(calculateDelay(attempt, retryConfig.RetryDelay, retryConfig.MaxDelay, retryConfig.RetryDelayJitter, retryConfig.BackOffFactor))
		}
	}

	return nil, httptransport.NewErrorResponse[E](errors.New("max retries exceeded"), nil)
}

// HandleStream processes a streaming request with automatic retry logic on retryable HTTP failures.
// Retries failed stream connections with exponential backoff.
func (h *RetryHandler[T, E]) HandleStream(request httptransport.Request) (*httptransport.Stream[T], *httptransport.ErrorResponse[E]) {
	if h.nextHandler == nil {
		return nil, httptransport.NewErrorResponse[E](errors.New("Handler chain terminated without terminating handler"), nil)
	}

	retryConfig := request.Config.Retry
	if retryConfig.MaxAttempts <= 0 {
		retryConfig.MaxAttempts = 1
	}
	for attempt := 0; attempt < retryConfig.MaxAttempts; attempt++ {
		stream, err := h.nextHandler.HandleStream(request.Clone())
		if err == nil {
			return stream, nil
		}

		if !h.shouldRetry(err, request.Method, retryConfig.HTTPMethodsToRetry, retryConfig.HTTPCodesToRetry) {
			return nil, err
		}

		if attempt == retryConfig.MaxAttempts-1 {
			return nil, err
		}

		if d, ok := retryAfterDelay(err.GetHeader(http.CanonicalHeaderKey("Retry-After-Ms")), err.GetHeader(http.CanonicalHeaderKey("Retry-After")), err.GetHeader(http.CanonicalHeaderKey("X-RateLimit-Reset")), retryConfig.MaxRetryAfterDelay); ok {
			time.Sleep(d)
		} else {
			time.Sleep(calculateDelay(attempt, retryConfig.RetryDelay, retryConfig.MaxDelay, retryConfig.RetryDelayJitter, retryConfig.BackOffFactor))
		}
	}

	return nil, httptransport.NewErrorResponse[E](errors.New("max retries exceeded"), nil)
}

// SetNext sets the next handler in the chain.
func (h *RetryHandler[T, E]) SetNext(handler Handler[T, E]) {
	h.nextHandler = handler
}
